Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::flattenDescendantLayersIfNecessary(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
+
+    // Layers are collected in depth first, post-order
+    Vector<TextureMapperLayer*> layersToFlatten;
+    collectLayersToFlattenRecursive(layersToFlatten);
+
+    for (auto* layer : layersToFlatten)
+        layer->flatten(options);
+}
+
+void TextureMapperLayer::collectLayersToFlattenRecursive(Vector<TextureMapperLayer*>& layers)
+{
+    // Traverse all children
+    for (auto* child : m_children) {
+        // Recursively collect from the child's descendants
+        child->collectLayersToFlattenRecursive(layers);
+
+        // If the child satisfies the condition, add it to the layers vector
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            layers.append(child);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Flattens descendant layers to into this planes z=0. In theory we could
+    // project each sublayer to this layer's plane but in practice that doesn't
+    // result in pixel perfect outcome and that's why each sublayer needs to be
+    // rendered into this layer's in 2D root.
```

Comment:
I don't think this comment is true.  In my understanding, all descendants of the flatten layer have to be rendered to z=0 of the flatten layer.

Reply:
What I tried to explain here is that instead of flattening subtree into separate 2D buffer we could try projecting child layers to flattening layer's plane in 3D coordinate space and thus avoid separate flattening buffer but in practice that does not result in pixel perfect outcome due to precision issues. I removed the comment to avoid confusion